sqlobject/src/main/java/org/jdbi/v3/sqlobject/config/internal/UseSqlParserImpl.java

```diff
@@ -26,43 +31,37 @@
         UseSqlParser anno = (UseSqlParser) annotation;
-        try {
-            final SqlParser parser = instantiate(anno.value(), sqlObjectType, method);
-            registry.get(SqlStatements.class).setSqlParser(parser);
-        } catch (Exception e) {
-            throw new IllegalStateException(e);
-        }
+        SqlParser parser = instantiate(anno.value(), sqlObjectType, method);
+        registry.get(SqlStatements.class).setSqlParser(parser);
     }
 
     @Override
     public void configureForType(ConfigRegistry registry, Annotation annotation, Class<?> sqlObjectType) {
-        UseSqlParser anno = (UseSqlParser) annotation;
-        try {
-            final SqlParser parser = instantiate(anno.value(), sqlObjectType, null);
-            registry.get(SqlStatements.class).setSqlParser(parser);
-        } catch (Exception e) {
-            throw new IllegalStateException(e);
-        }
+        configureForMethod(registry, annotation, sqlObjectType, null);
     }
 
-    private SqlParser instantiate(Class<? extends SqlParser> value,
+    private SqlParser instantiate(Class<? extends SqlParser> parserClass,
                                   Class<?> sqlObjectType,
-                                  Method m) throws Exception {
-        try {
-            Constructor<? extends SqlParser> noArg = value.getConstructor();
-            return noArg.newInstance();
-        } catch (NoSuchMethodException e) {
+                                  @Nullable Method method) {
+        return Stream.of(tryConstructor(parserClass), tryConstructor(parserClass, sqlObjectType), tryConstructor(parserClass, sqlObjectType, method))
+            .map(Supplier::get)
+            .filter(Objects::nonNull)
+            .findFirst()
+            .orElseThrow(() -> new IllegalStateException("Unable to instantiate, no viable constructor for " + parserClass.getName()));
+    }
+
+    private static <T extends SqlParser> Supplier<T> tryConstructor(Class<T> c, Object... args) {
+        Object[] nonNullArgs = Arrays.stream(args).filter(Objects::nonNull).toArray(Object[]::new);
+
+        return () -> {
             try {
-                Constructor<? extends SqlParser> classArg = value.getConstructor(Class.class);
-                return classArg.newInstance(sqlObjectType);
-            } catch (NoSuchMethodException e1) {
-                if (m != null) {
-                    Constructor<? extends SqlParser> constructor = value.getConstructor(Class.class,
-                                                                                            Method.class);
-                    return constructor.newInstance(sqlObjectType, m);
-                }
-                throw new IllegalStateException("Unable to instantiate, no viable constructor " + value.getName());
+                Class[] classes = Arrays.stream(nonNullArgs).map(Object::getClass).toArray(Class[]::new);
+                return c.getConstructor(classes).newInstance(nonNullArgs);
```

Comment:
If we're going to change this, we should change it to use the new `MethodHandles` style reflection.